Child conditions

treitter opened the pull request:
Adds support for child conditions, `start` and `stop` keywords and `parallel` blocks.

Commit:
Take advantage of dot notation for exclusive_conditions test

Commit:
Support subconditions and rule monitors

Subconditions (conditions which have at least one level of conditions
above them) automatically cause the creation of rule monitors when all
ancestor conditions are met.

While a monitor exists, the subcondition must be met by `start`
milliseconds and maintained until `stop` milliseconds after the creation
of the monitor. The `start` and `stop` times are specified in the rule
file by keywords of the same names.

If any ancestor condition becomes false, the monitor is cancelled
immediately.

This also adds support for 'parallel' blocks as a type of wrapper block.
Subconditions must be wrapped in a wrapper block.

Based on code originally written by Shane Fagan

Commit:
make is_string() safer

Commit:
Factor out keyword strings

Commit:
add TreeNode class to maintain hierarchy of rules file

This class allows maintaining the hierarchical structure of the config
file by tracking organization of the rules file elements in a tree
structure rather than lists.

araujol commented on sample_rules/exclusive_conditions.yaml:
```diff
@@ -26,7 +26,7 @@
     - condition: >
         lock_state == true &&
-        remote_key.command == 'lock'
+        remote_key == 'lock'
```
These changes are not required anymore since dot notation is already supported.

Actually it'd be better to change the identifier to use full dot notation: **remote.key.command**

araujol commented on sample_rules/monitored_condition.yaml:
```diff
@@ -0,0 +1,39 @@
+# We start the process by unconditionally emitting a signal, transmission_gear,
+# set to 'reverse'
+- emit:
+    signal: transmission_gear
```
It'd be much better to use dot notation for identifiers in new sample rules files.

araujol commented on vsm:
```diff
@@ -225,32 +318,87 @@ class State(object):
+        if isinstance(conditions, list) and isinstance(rules, dict):
+            for condition in conditions:
+                # TODO: replace with Parser
+                names, attributes = self.__get_ast_info(condition, names,
```
This should be replaced by a call to the **ParseIdentifiers** class.

araujol commented on vsm:
```diff
@@ -225,32 +318,87 @@ class State(object):
+
+    def add_rules(self, conditions, rules):
+        names = []
+        attributes = []
```
**attributes** is not used in this method?

araujol commented on vsm:
```diff
@@ -547,6 +843,8 @@ if __name__ == "__main__":
                 logger.e(err)
                 exit(1)
 
+        config_tree = TreeNode(NODE_ROOT, None)
```
I see this variable only being used in one place to add a child, what is exactly its function?

treitter replied:
It serves as the document root so we have a parent for the arbitrary number of top-level list items.

araujol commented:
@treitter Testing the _monitored_condition.yaml_ rules file, I increased the **start** value to 10000ms and **stop** value to 20000ms.

Then I set **transmission_gear** to true and just wait to receive a log error for the monitoring condition, and after 10 seconds, I get the following:

```
$ ./vsm ../monitored.yaml --log-file=-
< 10,transmission_gear,[SIGNUM],'reverse'
< 10,transmission_gear,[SIGNUM],'reverse'
transmission_gear = 'reverse'
> 1157,transmission_gear,[SIGNUM],"'reverse'"
> 1157,transmission_gear,[SIGNUM],'reverse'
< 1158,lights_external_backup,[SIGNUM],'True'
State = {
transmission_gear = reverse
}
< 1158,lights_external_backup,[SIGNUM],'True'
condition: (transmission_gear == 'reverse') => True
condition not met by 'stop' time of 20000ms
```

The log error shows:
```
condition not met by 'stop' time of 20000ms
```

Shouldn't that show instead an error about the condition not met by `start` time of 10000 ms?

treitter commented:
@araujol you're right about the error with the `start` timeout. It should be based on the `start` time, not the `stop` time. Fixed locally.

treitter commented:
@araujol: I've force-pushed changes based on your comments. Please re-review.

araujol commented:
@treitter The code looks better, and testing I can't reproduce my previous reported issue, but now it seems the log error is not recorded if the condition doesn't meet the **stop** time interval instead.

Changing values (like in the previous example) for **start:** 10000 , and **stop:** 20000 , I set `transmission.gear = 'reverse'` , and before the start time is reached, I set `camera.backup.active = true` , then wait until it goes beyond the **start** time and set  `camera.backup.active = false`, and keep this value until it goes beyond **stop** time, but it doesn't log any error.

Commit:
Avoid overwriting the global 'start_time_ms' in State class

The State class set the global 'start_time_ms' in 'handle_condition()' but
meant to only set a local variable.

treitter commented:
@araujol I was able to replicate locally, fixed the issue, and force-pushed. Please re-review.

I also added a trivial commit after this to prevent overwriting a global variable in `State.handle_condition()` though it doesn't appear it was yet causing any issues.

araujol commented on vsm:
```diff
@@ -225,32 +318,68 @@ class State(object):
+                if identifier not in self.signals:
+                    self.signals[identifier] = []
+
+                self.signals[name].append(rule_id)
```
This should be indexed by **identifier** instead of **name** ?

araujol commented on vsm:
```diff
@@ -225,32 +318,68 @@ class State(object):
+                if rules:
+                    self.add_rules(rules[0], rules[1])
+
+    def __exec_child_rule(self, rule):
```
@treitter Could you explain to me the exact purpose of this method?, I fail to see the effect it has.

araujol commented:
@treitter Nice, I can't reproduce the previous issue, it is working fine now.

I added a couple of comments about some details in the code, and I am also a bit concerned that none of the new tests are enabled for the IPC module, this means these new features won't be tested for the IPC plugins?

treitter commented:
I've addressed the latest comments in the new version of the branch I just force-pushed (in part by deleting code left over from the original branch which apparently didn't matter to the functionality of the tests).

And I've opened another task to enable more of the tests to run under IPC (project time permitting).

araujol commented on vsm:
```diff
@@ -241,16 +340,18 @@ class State(object):
         # No conditions based on the signal that was emitted,
         # nothing to be done.
-        if not signal in self.rules:
+        if not signal in self.rules and not signal in self.signals:
```
**self.signals** is not defined in this class.

araujol commented:
@treitter Just added a last comment about a tiny detail, apart of this, this code looks good to merge.

treitter commented:
@araujol This merged code addresses your last point as well
